orchagent/tunneldecaporch.cpp

```diff
@@ -249,44 +287,420 @@ void TunnelDecapOrch::doTask(Consumer& consumer)
+                {
+                    set_src_ip = true;
+                    src_ip_str = fvValue(fv);
+                    try
+                    {
+                        src_ip = IpPrefix(fvValue(fv));
+                    }
+                    catch (const std::invalid_argument &e)
+                    {
+                        SWSS_LOG_ERROR("%s: invalid source IP prefix %s.", key.c_str(), src_ip_str.c_str());
+                        valid = false;
+                        break;
+                    }
+                }
+                else if (fvField(fv) == "term_type")
+                {
+                    auto it = DecapTermTypes.find(fvValue(fv));
+                    if (it == DecapTermTypes.end())
+                    {
+                        SWSS_LOG_ERROR("%s: invalid tunnel decap term type %s.", key.c_str(), fvValue(fv).c_str());
+                        valid = false;
+                        break;
+                    }
+                    term_type = it->second;
+                }
+                else if (fvField(fv) == "subnet_type")
+                {
+                    subnet_type = fvValue(fv);
+                    if (subnet_type != "vlan" && subnet_type != "vip")
+                    {
+                        SWSS_LOG_ERROR("%s: invalid subnet type: %s.", key.c_str(), subnet_type.c_str());
+                        valid = false;
+                        break;
+                    }
+                }
+                else
+                {
+                    SWSS_LOG_ERROR("%s: unknown decap term table attribute '%s'", key.c_str(), fvField(fv).c_str());
+                    valid = false;
+                    break;
+                }
+            }
+
+            if (valid)
+            {
+                if ((is_subnet_decap_v4_term || is_subnet_decap_v6_term) && term_type != TUNNEL_TERM_TYPE_MP2MP)
+                {
+                    SWSS_LOG_ERROR("%s: only MP2MP tunnel decap term is allowed for subnet decap tunnel.", key.c_str());
+                    valid = false;
+                }
+                else if (term_type == TUNNEL_TERM_TYPE_P2MP && set_src_ip)
+                {
+                    SWSS_LOG_ERROR("%s: source IP is not allowed for P2MP tunnel decap term.", key.c_str());
+                    valid = false;
+                }
+                else if (!subnet_type.empty() && term_type != TUNNEL_TERM_TYPE_MP2MP)
+                {
+                    SWSS_LOG_ERROR("%s: only MP2MP is allowed for subnet decap term.", key.c_str());
+                    valid = false;
+                }
+                else if (term_type == TUNNEL_TERM_TYPE_P2P && !set_src_ip)
+                {
+                    SWSS_LOG_ERROR("%s: no source IP is provided.", key.c_str());
+                    valid = false;
+                }
+                else if (term_type == TUNNEL_TERM_TYPE_MP2MP && !is_subnet_decap_v4_term && !is_subnet_decap_v6_term && !set_src_ip)
+                {
```

Comment:
Is it necessary to check `!is_subnet_decap_v4_term && !is_subnet_decap_v6_term` ?

Reply:
Check below.